src/java/org/apache/cassandra/db/ReadCommand.java

```diff
@@ -632,6 +654,84 @@ private void maybeDelayForTesting()
         }
     }
 
+    private UnfilteredPartitionIterator withQuerySizeTracking(UnfilteredPartitionIterator iterator)
+    {
+        if (!trackWarnings || SchemaConstants.isSystemKeyspace(metadata().keyspace)) // exclude internal keyspaces
```

Comment:
Say we have RF=3, and we're always issuing a speculative read during a QUORUM query. If the local node is providing a digest response, it is possible for there to be a mismatch when the other two replicas respond, and then for the local node to, on a separate thread, execute a read-repair-driven data read. In that case, are we guaranteed that `trackWarnings` will be properly visible to that other thread (a `ReadStage` thread)?

Generally, I want to make sure we have our heads wrapped around the paths in the code (read-repair, short read protection, replica filtering protection) that hit `Stage.READ.maybeExecuteImmediately()` and can transfer execution to `ReadStage` threads while mutable state is perhaps being relied on in `ReadCommand`. For instance, the approach CASSANDRA-16721 took around this was moving as much as possible relating to a local read through the local read runnable itself, which, once it starts executing, isn't going to randomly switch threads.

Reply:
Take a look at `RepairDigestTrackingTest#testLocalDataAndRemoteRequestConcurrency()`.

Reply:
>  are we guaranteed that trackWarnings will be properly visible to that other thread (a ReadStage thread)?

My gut tells me there will be a bug but won't know till I test.  Will look closer at your test and try to replicate

The bigger question is will we propagate the warnings, and I feel like we won't (it was annoying to get scans working... and that is single threaded!).  If we cross thread-boundaries I believe we will load information, so the repair will act as if track_warnings was disabled (even if the field is correct as I made sure cloning propagates the field)

Reply:
in a debugger I do see the following codepath called for a digest request and a read repair

```
// We delay the local (potentially blocking) read till the end to avoid stalling remote requests.
        if (hasLocalEndpoint)
        {
            logger.trace("reading {} locally", readCommand.isDigestQuery() ? "digest" : "data");
            Stage.READ.maybeExecuteImmediately(new LocalReadRunnable(readCommand, handler));
        }
```

this makes sense why this doesn't have any issues, as the reply to the callback is a Message and we inject the result into the Message; so the only issue should be on the visibility of the field; so need to look at access

![Screen Shot 2021-09-07 at 4 32 18 PM](https://user-images.githubusercontent.com/1852601/132423123-286e863c-e241-4c70-a8b4-664df16608e6.png)

Here we see digests are cloned, and in doing so we set this field in the constructor (though not final); example

```
protected SinglePartitionReadCommand copyAsDigestQuery()
    {
        return new SinglePartitionReadCommand(true,
                                              digestVersion(),
                                              acceptsTransient(),
                                              metadata(),
                                              nowInSec(),
                                              columnFilter(),
                                              rowFilter(),
                                              limits(),
                                              partitionKey(),
                                              clusteringIndexFilter(),
                                              indexMetadata(),
                                              isTrackingWarnings());
    }
```

the only other path is at the start right after making the ReadCommand.

Given this, it doesn't seem that there are concurrency issues, though there is an issue where we report warn multiple times (this seems to be an issue with ClientWarn in general)

Reply:
here are the call paths I see which call awaitResponse on ReadCallback (which adds the warnings)

```
org.apache.cassandra.service.reads.ReadCallback#awaitResults
org.apache.cassandra.service.reads.AbstractReadExecutor#awaitResponses
org.apache.cassandra.service.StorageProxy#fetchRows
org.apache.cassandra.service.StorageProxy#readRegular
org.apache.cassandra.service.StorageProxy#read
org.apache.cassandra.db.SinglePartitionReadCommand$Group#execute
org.apache.cassandra.cql3.statements.SelectStatement#execute
org.apache.cassandra.cql3.statements.SelectStatement#execute
org.apache.cassandra.cql3.statements.SelectStatement#execute
```

```
org.apache.cassandra.service.reads.ReadCallback#awaitResults
org.apache.cassandra.service.reads.repair.AbstractReadRepair#awaitReads
org.apache.cassandra.service.reads.AbstractReadExecutor#awaitReadRepair
org.apache.cassandra.service.StorageProxy#fetchRows
org.apache.cassandra.service.StorageProxy#readRegular
org.apache.cassandra.service.StorageProxy#read
org.apache.cassandra.db.SinglePartitionReadCommand$Group#execute
org.apache.cassandra.cql3.statements.SelectStatement#execute
org.apache.cassandra.cql3.statements.SelectStatement#execute
org.apache.cassandra.cql3.statements.SelectStatement#execute
```

Based off code inspection I also see short-read-protection would have the same issue.

Reply:
trying to read what's going on in `read-repair`, `short read protection`, and `replica filtering protection` to see what is the best place to dedup.  So far read-repair looks like you need to handle in `AbstractReadExecutor`, but need to read the other two closer to confirm...

Reply:
took a while but ended up fixing with a thread-local...  We now log once per ReadCommand (cloning will cause extra logging, also changing the query (aka paging))

Reply:
Resolving as we now test that RR does not trigger extra logging